sdk/src/azure/iot/az_iot_pnp_client_twin.c

```diff
@@ -0,0 +1,379 @@
+    _az_RETURN_IF_FAILED(az_json_writer_append_string(json_writer, component_property_label_value));
+  }
+
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(json_writer, property_name));
+  _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(json_writer));
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(json_writer, property_ack_code_name));
+  _az_RETURN_IF_FAILED(az_json_writer_append_int32(json_writer, ack_code));
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(json_writer, property_ack_version_name));
+  _az_RETURN_IF_FAILED(az_json_writer_append_int32(json_writer, ack_version));
+
+  if (az_span_size(ack_description) != 0)
+  {
+    _az_RETURN_IF_FAILED(
+        az_json_writer_append_property_name(json_writer, property_ack_description_name));
+    _az_RETURN_IF_FAILED(az_json_writer_append_string(json_writer, ack_description));
+  }
+
+  _az_RETURN_IF_FAILED(
+      az_json_writer_append_property_name(json_writer, property_response_value_name));
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_pnp_client_twin_end_property_with_status(
+    az_iot_pnp_client const* client,
+    az_json_writer* json_writer,
+    az_span component_name)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(json_writer);
+
+  (void)client;
+
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(json_writer));
+
+  if (az_span_size(component_name) != 0)
+  {
+    _az_RETURN_IF_FAILED(az_json_writer_append_end_object(json_writer));
+  }
+
+  return AZ_OK;
+}
+
+// Move reader to the value of property name
+static az_result json_child_token_move(az_json_reader* jr, az_span property_name)
+{
+  do
+  {
+    if ((jr->token.kind == AZ_JSON_TOKEN_PROPERTY_NAME)
+        && az_json_token_is_text_equal(&(jr->token), property_name))
+    {
+      if (az_result_failed(az_json_reader_next_token(jr)))
+      {
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      return AZ_OK;
+    }
+    else if (jr->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
+    {
+      if (az_result_failed(az_json_reader_skip_children(jr)))
+      {
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+    }
+    else if (jr->token.kind == AZ_JSON_TOKEN_END_OBJECT)
+    {
+      return AZ_ERROR_ITEM_NOT_FOUND;
+    }
+  } while (az_result_succeeded(az_json_reader_next_token(jr)));
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+// Check if the component name is in the model
+static az_result is_component_in_model(
+    az_iot_pnp_client const* client,
+    az_json_token* component_name,
+    az_span* out_component_name)
+{
+  int32_t index = 0;
+
+  while (index < client->_internal.options.component_names_length)
+  {
+    if (az_json_token_is_text_equal(
+            component_name, client->_internal.options.component_names[index]))
+    {
+      *out_component_name = client->_internal.options.component_names[index];
+      return AZ_OK;
+    }
+
+    index++;
+  }
+
+  return AZ_ERROR_UNEXPECTED_CHAR;
```

Comment:
I don't know if this is the right error for this method. Change it to return `bool`. It is an `is_...` method, after all.